### PR TITLE
Use scalajs-react's dispatcher for async effects/useDebouncedTimeout hook/{set,mod}StateAsync for useState

### PR DESCRIPTION
**Pull request by rpiaggio**
Should probably have been 3 separate PRs.

* Use `scalajs-react`'s dispatcher for async effects - When using `IO` as async effect, `scalajs-react` keeps an instance of `Dispatcher`. By switching to it (via the `Dispatch` typeclass), we avoid having to pass an implicit one (in explore this means we don't need the implicit context, where we store the `Dispatcher[IO]`).

* `useDebouncedTimeout` hook - Useful for example when we want to trigger an effect some time after the user stops typing in an `input`.

* `{set,mod}StateAsync` - Convenience methods to run state modifications in the default async effect instead of the sync one. We already had this for state in class components, now we have them for state obtained through `useState` hook.

### CODE REVIEW

File: js/src/main/scala/crystal/react/hooks/TimeoutHandle.scala [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import cats.effect.Temporal
+import cats.syntax.all._
+import crystal.react.implicits._
+import japgolly.scalajs.react.hooks.Hooks
+import japgolly.scalajs.react.util.DefaultEffects.{ Async => DefaultA }
+import japgolly.scalajs.react.util.DefaultEffects.{ Sync => DefaultS }
+
+import scala.concurrent.duration.FiniteDuration
+
+class TimeoutHandle(
```
Comment (cquiroz) [inline js/src/main/scala/crystal/react/hooks/TimeoutHandle.scala:13]: Cool

File: js/src/main/scala/crystal/react/hooks/UseDebouncedTimeout.scala [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  */
+object UseDebouncedTimeout {
+
+  // Changing duration while component is mounted is not supported.
+  val hook = CustomHook[FiniteDuration]
+    .useRef(none[Fiber[DefaultA, Throwable, Unit]])
+    .useMemoBy((_, _) => ())((duration, timerRef) => _ => new TimeoutHandle(duration, timerRef))
+    .buildReturning((_, _, timeoutHandle) => timeoutHandle)
+
+  object HooksApiExt {
```
Comment (cquiroz) [inline js/src/main/scala/crystal/react/hooks/UseDebouncedTimeout.scala:25]: I guess this is you custom hook?

File: project/plugins.sbt [hunk 3]
```diff
@@ -3,3 +3,4 @@ addSbtPlugin("org.portable-scala"        % "sbt-scalajs-crossproject" % "1.1.0")
 addSbtPlugin("com.github.sbt"            % "sbt-ci-release"           % "1.5.10")
 addSbtPlugin("org.scalameta"             % "sbt-scalafmt"             % "2.4.3")
 addSbtPlugin("io.github.davidgregory084" % "sbt-tpolecat"             % "0.1.20")
+addSbtPlugin("ch.epfl.scala"             % "sbt-scalafix"             % "0.9.31")
```
Comment (cquiroz) [inline project/plugins.sbt:6]: There is a new version 0.9.32

**Review by cquiroz — APPROVED** [review]
LGTM

File: project/plugins.sbt (hunk 3 above)
Comment (rpiaggio) [reply project/plugins.sbt:6]: thx :+1:

File: js/src/main/scala/crystal/react/hooks/UseDebouncedTimeout.scala (hunk 2 above)
Comment (rpiaggio) [reply js/src/main/scala/crystal/react/hooks/UseDebouncedTimeout.scala:25]: Right, this is to have nice syntax `ScalaFnComponent.withHooks[Props].useOurCustomHook`

**Pull request merged**